Field constraints

esteve opened the pull request:
Connects to #67

esteve commented:
CI:

http://ci.ros2.org/job/ci_linux/843/
http://ci.ros2.org/job/ci_osx/706/
http://ci.ros2.org/job/ci_windows/915/

esteve commented:
Question:

Should we be strict about checking types or should we use `isinstance`? Also, in the case of arrays, should we accept `tuples` and other `iterable` types (except strings) as well? This should go into the article too, but I prefer arguing here with actual code and then paste the conclusion in the article.

wjwwood commented:
> Should we be strict about checking types or should we use `isinstance`?

Do you mean should we allow subclasses? Maybe you could expand on the two options, as in what would be possible with one versus the other?

> Also, in the case of arrays, should we accept `tuple`s and other `iterable` types (except `string`s) as well?

I think `tuple`s would be ok, but `dict` doesn't really make sense for example.

esteve commented:
> Do you mean should we allow subclasses? Maybe you could expand on the two options, as in what would be possible with one versus the other?

Yes, I meant that. I'd prefer not supporting subclasses, but @dirk-thomas argued in https://github.com/ros2/rosidl/pull/63 that's a bit too restrictive for checking equality, so I don't know if that applies here too.

> I think tuples would be ok, but dict doesn't really make sense for example.

Yes. How about custom defined `iterable`s? I'd rather restrict it to `tuple`s and `list`s only, though.

esteve commented:
I've updated this PR to support `tuple`s for static arrays, but not for dynamic arrays.

wjwwood commented:
I don't see that there is any extra work that needs to be done to support subclassing, nor a good reason to disallow it. Can you provide some justification for avoiding it?

It be nice to be able to assign a tuple to any kind of "array" in the message, even if it just implicitly converts it to a list. It might be worth considering how `np.array` could be used as well, because sometimes converting on assignment might be unnecessarily expensive. So I might go back on my previous statement and agree that any iterable should be allowed so long as it contains the right types.

Also, this got me thinking that you'll need to check the types on publish too. For example, a reasonable error when someone does this would be nice:

```python
msg = Foo()
"""
# Foo.msg
float64[] floats
"""

msg.floats.append(1.3)
msg.floats.append(object())
msg.floats.append(None)
publisher.publish(msg)  # Should get a reasonable error here.
```

dirk-thomas commented:
Since Pythons philosophy is based on dock typing I think we should follow it. Anything which can 
"look" like the expected data should be allowed. I don't think we should restrict this in any way. As long as we can serialize / deserialize the data into the format specified in the rosidl it should be ok for the user to use that value.

esteve commented:
One way is, for tuples and lists (and similar objects), to just check that they implement the sequence protocol. Unfortunately, numpy's ndarray does not implement it. Any idea of how to generalize such logic or should we special-case for numpy?

esteve commented:
I just rebased to resolve the conflicts.

esteve commented:
I've replaced strict type checking with `isinstance`, checking the `Sequence` protocol instead for arrays. I've also replaced the other type checks (strings, bytes, etc.) with `isinstance`. I'm not checking for `numpy` arrays yet, though, as I'd rather start with the core language and add support for `numpy` later.

dirk-thomas commented on rosidl_generator_py/resource/_msg.py.template:
```diff
@@ -53,4 +53,21 @@ class @(spec.base_type.type)(metaclass=@(spec.base_type.type)__Constants):
+        from collections import ByteString
+@[end if]@
+@[if field.type.type in ['char', 'string']]@
+        from collections import UserString
```
Will this result in a duplicate import together with line 63?

esteve replied:
Good catch, fixing it now.

dirk-thomas commented on rosidl_generator_py/resource/_msg.py.template:
```diff
@@ -53,4 +53,21 @@ class @(spec.base_type.type)(metaclass=@(spec.base_type.type)__Constants):
+@[if field.type.type in ['char', 'string']]@
+        from collections import UserString
+@[end if]@
+        @(check_constraints(field.type))
```
Instead of constructing this code in Python can this be done in the template instead? The tests should then not test the function but the expected behavior of the generated code.

esteve replied:
I started moving the code into the Python counterpart, but yeah, over time it all looks very convoluted now. I'll move it back to the template.

esteve commented:
CI:

http://ci.ros2.org/job/ci_linux/854/
http://ci.ros2.org/job/ci_osx/716/
http://ci.ros2.org/job/ci_windows/937/

dirk-thomas commented:
lgtm

esteve commented:
I've squashed the commits and put all the quotes replacements (from double to single) in a separate commit.

CI:

http://ci.ros2.org/job/ci_linux/855/
http://ci.ros2.org/job/ci_osx/717/
http://ci.ros2.org/job/ci_windows/938/

esteve commented:
I'm going to merge this. I rewrote the history of the commits, but nothing was added/removed.
